src/sinks/humio_logs.rs

```diff
@@ -118,3 +118,156 @@ mod tests {
         assert_eq!((hec_event.time * 1000f64).fract(), 0f64);
     }
 }
+
+#[cfg(test)]
+#[cfg(feature = "humio-integration-tests")]
+mod integration_tests {
+    use super::*;
+    use crate::{
+        sinks::util::Compression,
+        test_util::{random_string, runtime},
+        topology::config::{SinkConfig, SinkContext},
+        Event,
+    };
+    use futures::compat::Future01CompatExt;
+    use futures01::Sink;
+    use serde_json::json;
+    use serde_json::Value as JsonValue;
+
+    // matches humio container address
+    const HOST: &str = "http://localhost:8080";
+
+    #[test]
+    fn humio_insert_message() {
+        let mut rt = runtime();
+        let cx = SinkContext::new_test();
+
+        rt.block_on_std(async move {
```

Comment:
Can this be simplified with `#[tokio::test]`?

Reply:
I tried this, but it seems like the sink wasn't running when I went to `.send` to it.

```
thread 'sinks::humio_logs::integration_tests::humio_insert_message' panicked at 'Spawn service future: SpawnError { is_shutdown: true }', src/sinks/util/sink.rs:259:21
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: rust_begin_unwind
             at src/libstd/panicking.rs:388
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:101
  13: core::option::expect_none_failed
             at src/libcore/option.rs:1272
  14: core::result::Result<T,E>::expect
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/result.rs:963
  15: <vector::sinks::util::sink::BatchSink<S,B,Request,E> as futures::sink::Sink>::poll_complete
             at src/sinks/util/sink.rs:259
  16: <vector::sinks::util::sink::BatchSink<S,B,Request,E> as futures::sink::Sink>::start_send
             at src/sinks/util/sink.rs:232
  17: <vector::sinks::util::http::BatchedHttpSink<T,B,L> as futures::sink::Sink>::poll_complete
             at src/sinks/util/http.rs:164
  18: <vector::sinks::util::http::BatchedHttpSink<T,B,L> as futures::sink::Sink>::start_send
             at src/sinks/util/http.rs:156
  19: <futures::sink::map_err::SinkMapErr<S,F> as futures::sink::Sink>::start_send
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/sink/map_err.rs:45
  20: <alloc::boxed::Box<S> as futures::sink::Sink>::start_send
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/sink/mod.rs:64
  21: <futures::sink::send::Send<S> as futures::future::Future>::poll
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/sink/send.rs:46
  22: core::ops::function::FnOnce::call_once
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232
  23: futures::task_impl::Spawn<T>::enter::{{closure}}
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/task_impl/mod.rs:399
  24: futures::task_impl::std::set
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/task_impl/std/mod.rs:83
  25: futures::task_impl::Spawn<T>::enter
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/task_impl/mod.rs:399
  26: futures::task_impl::Spawn<T>::poll_fn_notify
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.29/src/task_impl/mod.rs:291
  27: futures_util::compat::compat01as03::Compat01As03<T>::in_notify
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/compat/compat01as03.rs:42
  28: <futures_util::compat::compat01as03::Compat01As03<Fut> as core::future::future::Future>::poll
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/compat/compat01as03.rs:162
  29: vector::sinks::humio_logs::integration_tests::humio_insert_message::{{closure}}
             at src/sinks/humio_logs.rs:152
  30: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/future/mod.rs:73
  31: tokio::runtime::enter::Enter::block_on::{{closure}}
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/enter.rs:163
  32: tokio::coop::with_budget::{{closure}}
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/coop.rs:127
  33: std::thread::local::LocalKey<T>::try_with
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/local.rs:263
  34: std::thread::local::LocalKey<T>::with
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/thread/local.rs:239
  35: tokio::coop::with_budget
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/coop.rs:120
  36: tokio::coop::budget
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/coop.rs:96
  37: tokio::runtime::enter::Enter::block_on
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/enter.rs:163
  38: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/thread_pool/mod.rs:82
  39: tokio::runtime::Runtime::block_on::{{closure}}
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/mod.rs:446
  40: tokio::runtime::context::enter
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/context.rs:72
  41: tokio::runtime::handle::Handle::enter
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/handle.rs:76
  42: tokio::runtime::Runtime::block_on
             at /home/jesse/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.21/src/runtime/mod.rs:441
  43: vector::sinks::humio_logs::integration_tests::humio_insert_message
             at src/sinks/humio_logs.rs:141
  44: vector::sinks::humio_logs::integration_tests::humio_insert_message::{{closure}}
             at src/sinks/humio_logs.rs:141
  45: core::ops::function::FnOnce::call_once
             at /home/jesse/.rustup/toolchains/1.45.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ops/function.rs:232
  46: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/liballoc/boxed.rs:1076
  47: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panic.rs:318
  48: std::panicking::try::do_call
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:297
  49: std::panicking::try
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:274
  50: std::panic::catch_unwind
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panic.rs:394
  51: test::run_test_in_process
             at src/libtest/lib.rs:541
  52: test::run_test::run_test_inner::{{closure}}
             at src/libtest/lib.rs:450
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test sinks::humio_logs::integration_tests::humio_insert_message ... FAILED

failures:

failures:
    sinks::humio_logs::integration_tests::humio_insert_message
```

Luke suggested that it might be because this sink has some dependency on tokio01 and could be running on a different background.

Reply:
Ok, it was a question based on seeing that pattern starting to be useful in other tests. I think we will likely move there eventually, but it seems we'll have to wait until we're rid of all the tokio01 bits.